lib/services/role.go

```diff
@@ -1428,141 +1416,119 @@ func (set RoleSet) CheckAccessToKubernetes(namespace string, kube *KubernetesClu
+type DatabaseLabelsMatcher struct {
+	Labels map[string]string
+}
+
+// Match matches database server labels against provided role and condition.
+func (m *DatabaseLabelsMatcher) Match(role Role, condition RoleConditionType) (bool, error) {
+	match, _, err := MatchLabels(role.GetDatabaseLabels(condition), m.Labels)
+	return match, trace.Wrap(err)
+}
+
+// DatabaseUserMatcher matches database account name.
+type DatabaseUserMatcher struct {
+	User string
+}
+
+// Match matches database account name against provided role and condition.
+func (m *DatabaseUserMatcher) Match(role Role, condition RoleConditionType) (bool, error) {
+	match, _ := MatchDatabaseUser(role.GetDatabaseUsers(condition), m.User)
+	return match, nil
+}
+
+// DatabaseNameMatcher matches database name.
+type DatabaseNameMatcher struct {
+	Name string
+}
+
+// Match matches database name against provided role and condition.
+func (m *DatabaseNameMatcher) Match(role Role, condition RoleConditionType) (bool, error) {
+	match, _ := MatchDatabaseName(role.GetDatabaseNames(condition), m.Name)
+	return match, nil
 }
 
-// CheckAccessToDatabaseUser checks if this role set is allowed to connect to
-// a database as a particular database account.
+// CheckAccessToDatabase checks if this role set has access to a particular database.
 //
-// Used as an authorization check when connecting to a database.
-func (set RoleSet) CheckAccessToDatabaseUser(server types.DatabaseServer, dbUser string) error {
-	var errs []error
+// The checker always checks the server namespace, other matches are supplied
+// by the caller.
+func (set RoleSet) CheckAccessToDatabase(server types.DatabaseServer, matchers ...Matcher) error {
+	log := log.WithField(trace.Component, teleport.ComponentRBAC)
 	// Check deny rules.
 	for _, role := range set {
-		matchNamespace, namespaceMessage := MatchNamespace(role.GetNamespaces(Deny), server.GetNamespace())
-		matchLabels, labelsMessage, err := MatchLabels(role.GetDatabaseLabels(Deny), server.GetAllLabels())
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		matchUser, userMessage := MatchDatabaseUser(role.GetDatabaseUsers(Deny), dbUser)
-		if matchNamespace && matchLabels && matchUser {
-			log.WithField(trace.Component, teleport.ComponentRBAC).Debugf(
-				"Access to database %q (dbuser=%v) denied, deny rule in %q matched; match(namespace=%v, label=%v, dbuser=%v).",
-				server.GetName(), dbUser, role.GetName(), namespaceMessage, labelsMessage, userMessage)
-			return trace.AccessDenied("access to database denied")
+		matchNamespace, _ := MatchNamespace(role.GetNamespaces(Deny), server.GetNamespace())
+		// Deny rules are greedy on purpose. They will always match if
+		// at least one of the matchers returns true.
+		if matchNamespace {
+			match, matcher, err := Matchers(matchers).MatchAny(role, Deny)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			if match {
+				log.Debugf("Access to database %q denied, deny rule in role %q matched; %#v.",
```

Comment:
I'd suggest that matchers implement fmt.Stringer instead of dumping them raw to the logs.